### PR TITLE
Load Posts from Server 

### PR DESCRIPTION
**Pull request by mvlexs**
solved issue #1

### CODE REVIEW

File: app/build.gradle.kts [hunk 1]
```diff
@@ -100,6 +101,7 @@ kotlin {
 }
 
 dependencies {
+    implementation(libs.firebase.crashlytics.buildtools)
```
Comment (jp-hoehmann) [inline app/build.gradle.kts]: Remove this

File: app/src/main/java/com/example/jetnews/data/posts/impl/FakePostsRepository.kt
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/data/posts/impl/FakePostsRepository.kt:1]: If this is no longer needed, remove this file

File: app/src/main/java/com/example/jetnews/data/posts/impl/HttpPostsRepository.kt [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    // used to drive "random" failure in a predictable pattern, making the first request always
+    // succeed
+    private var requestCount = 0
+
+    /**
+     * Randomly fail some loads to simulate a real network.
+     *
+     * This will fail deterministically every 5 requests
+     */
+    private fun shouldRandomlyFail(): Boolean = ++requestCount % 5 == 0
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/data/posts/impl/HttpPostsRepository.kt]: Remove this

File: app/src/main/java/com/example/jetnews/data/posts/impl/HttpPostsRepository.kt [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+import com.example.jetnews.model.Post
+import com.example.jetnews.model.PostsFeed
+import com.example.jetnews.utils.addOrRemove
+import io.ktor.client.HttpClient
+import io.ktor.client.call.body
+import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
+import io.ktor.client.request.get
+import io.ktor.serialization.kotlinx.json.json
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.delay
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/data/posts/impl/HttpPostsRepository.kt]: Remove this

File: app/src/main/java/com/example/jetnews/data/posts/impl/PostsData.kt [hunk 4]
```diff
@@ -17,6 +17,13 @@
 @file:Suppress("ktlint:max-line-length") // String constants read better
 package com.example.jetnews.data.posts.impl
 
+import android.util.Log
+import androidx.glance.LocalContext
+import com.android.volley.BuildConfig
+import com.android.volley.Request
+import com.android.volley.Response
+import com.android.volley.toolbox.JsonObjectRequest
+import com.example.jetnews.JetnewsApplication
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/data/posts/impl/PostsData.kt]: Remove this

File: app/src/main/java/com/example/jetnews/data/posts/impl/PostsData.kt [hunk 5]
```diff
@@ -27,6 +34,14 @@ import com.example.jetnews.model.Post
...
 import com.example.jetnews.model.PostsFeed
 import com.example.jetnews.model.Publication
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.json.JsonElement
+import org.json.JSONObject
+import java.net.URL
+
+
+
+
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/data/posts/impl/PostsData.kt]: Remove this

File: app/src/main/java/com/example/jetnews/data/posts/impl/PostsData.kt [hunk 6]
```diff
@@ -322,7 +337,7 @@ val paragraphsPost2 = listOf(
     Paragraph(
         ParagraphType.CodeBlock,
         """
-        @Inject @MinimumBalance lateinit var minimumBalance: BigDecimal 
+        @Inject @MinimumBalance lateinit var minimumBalance: BigDecimal
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/data/posts/impl/PostsData.kt]: Revert unnecessary whitespace change

File: app/src/main/java/com/example/jetnews/data/posts/impl/PostsData.kt [hunk 7]
```diff
@@ -415,8 +430,8 @@ val paragraphsPost2 = listOf(
         """
         class MyVMFactory @Inject constructor(
           private val vmMap: Map<String, @JvmSuppressWildcards Provider<ViewModel>>
-        ) { 
-            ... 
+        ) {
+            ...
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/data/posts/impl/PostsData.kt]: Revert unecessary whitespace change

File: app/src/main/java/com/example/jetnews/JetnewsApplication.kt
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/JetnewsApplication.kt:1]: Revert all changes to this file

File: app/src/main/java/com/example/jetnews/glance/ui/Post.kt [hunk 8]
```diff
@@ -229,4 +271,4 @@ fun PostDescription(title: String, metadata: String, modifier: GlanceModifier) {
                 .copy(color = GlanceTheme.colors.onBackground)
         )
     }
-}
+}
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/glance/ui/Post.kt]: Revert unnecessary whitespace change

File: app/src/main/java/com/example/jetnews/model/Post.kt [hunk 9]
```diff
@@ -26,53 +30,75 @@ data class Post(
...
+    @SerialName("text")
     Text,
+    @SerialName("code-block")
     CodeBlock,
+    @SerialName("quote")
     Quote,
+    @SerialName("bullet")
     Bullet,
-}
+}
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/model/Post.kt]: Revert unnecessary whitespace change

File: app/src/main/java/com/example/jetnews/ui/article/ArticleScreen.kt [hunk 10]
```diff
@@ -58,8 +58,9 @@ import androidx.compose.ui.tooling.preview.Preview
 import androidx.compose.ui.unit.dp
 import com.example.jetnews.R
 import com.example.jetnews.data.Result
-import com.example.jetnews.data.posts.impl.BlockingFakePostsRepository
+import com.example.jetnews.data.posts.impl.HttpPostsRepository
 import com.example.jetnews.data.posts.impl.post3
+//import com.example.jetnews.data.posts.impl.post3
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/ui/article/ArticleScreen.kt]: Remove this

File: app/src/main/java/com/example/jetnews/ui/article/PostContent.kt [hunk 11]
```diff
@@ -68,8 +68,10 @@ import androidx.compose.ui.tooling.preview.Preview
 import androidx.compose.ui.unit.Dp
 import androidx.compose.ui.unit.dp
 import androidx.compose.ui.unit.sp
+import coil.compose.AsyncImage
 import com.example.jetnews.R
 import com.example.jetnews.data.posts.impl.post3
+//import com.example.jetnews.data.posts.impl.post3
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/ui/article/PostContent.kt]: Remove this

File: app/src/main/java/com/example/jetnews/ui/article/PostContent.kt [hunk 12]
```diff
@@ -117,9 +119,9 @@ private fun PostHeaderImage(post: Post) {
         .heightIn(min = 180.dp)
         .fillMaxWidth()
         .clip(shape = MaterialTheme.shapes.large)
-    Image(
-        painter = painterResource(post.imageId),
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/ui/article/PostContent.kt:121]: Remove import for painterResource

File: app/src/main/java/com/example/jetnews/ui/home/PostCardTop.kt [hunk 13]
```diff
@@ -53,9 +54,9 @@ fun PostCardTop(post: Post, modifier: Modifier = Modifier) {
             .heightIn(min = 180.dp)
             .fillMaxWidth()
             .clip(shape = MaterialTheme.shapes.large)
-        Image(
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/ui/home/PostCardTop.kt:56]: Remove unused import for Image

File: app/src/main/java/com/example/jetnews/ui/home/PostCardTop.kt [hunk 14]
```diff
@@ -53,9 +54,9 @@ fun PostCardTop(post: Post, modifier: Modifier = Modifier) {
...
-            painter = painterResource(post.imageId),
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/ui/home/PostCardTop.kt:57]: Remove unused import for painterResource

File: app/src/main/java/com/example/jetnews/ui/home/PostCardYourNetwork.kt [hunk 15]
```diff
@@ -63,15 +64,14 @@ fun PostCardPopular(
             .width(280.dp)
     ) {
         Column {
-            Image(
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/ui/home/PostCardYourNetwork.kt:66]: Remove unused import for Image

File: app/src/main/java/com/example/jetnews/ui/home/PostCardYourNetwork.kt [hunk 16]
```diff
@@ -63,15 +64,14 @@ fun PostCardPopular(
...
-                painter = painterResource(post.imageId),
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/ui/home/PostCardYourNetwork.kt:67]: Remove unused import for painterResource

File: app/src/main/java/com/example/jetnews/ui/home/PostCardYourNetwork.kt [hunk 17]
```diff
@@ -63,15 +64,14 @@ fun PostCardPopular(
...
-                contentDescription = null, // decorative
+            AsyncImage(
+                model = post.imageUrl,
+                contentDescription = null,
                 contentScale = ContentScale.Crop,
                 modifier = Modifier
                     .height(100.dp)
                     .fillMaxWidth()
             )
-
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/ui/home/PostCardYourNetwork.kt:74]: Revert unnecessary whitespace change

File: app/src/main/java/com/example/jetnews/ui/home/PostCardYourNetwork.kt [hunk 18]
```diff
@@ -114,7 +114,6 @@ fun PreviewPostCardPopular(
         }
     }
 }
-
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/ui/home/PostCardYourNetwork.kt]: Revert unnecessary whitespace change

File: app/src/main/java/com/example/jetnews/ui/home/PostCards.kt [hunk 19]
```diff
@@ -220,23 +223,23 @@ fun BookmarkButtonBookmarkedPreview() {
...
+//
+//@Preview("Post History card")
+//@Composable
+//fun HistoryPostPreview() {
+//    JetnewsTheme {
+//        Surface {
+//            PostCardHistory(post3, {})
+//        }
+//    }
+//}
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/ui/home/PostCards.kt]: Put these previews back

File: app/src/main/java/com/example/jetnews/ui/home/PostCards.kt [hunk 20]
```diff
@@ -21,6 +21,8 @@ import androidx.compose.foundation.Image
 import androidx.compose.foundation.clickable
 import androidx.compose.foundation.layout.Column
 import androidx.compose.foundation.layout.Row
+import androidx.compose.foundation.layout.fillMaxWidth
+import androidx.compose.foundation.layout.height
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/ui/home/PostCards.kt]: Remove this

File: app/src/main/java/com/example/jetnews/ui/home/PostCards.kt [hunk 21]
```diff
@@ -38,6 +40,7 @@ import androidx.compose.runtime.remember
 import androidx.compose.runtime.setValue
 import androidx.compose.ui.Modifier
 import androidx.compose.ui.draw.clip
+import androidx.compose.ui.layout.ContentScale
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/ui/home/PostCards.kt]: Remove this

File: app/src/main/java/com/example/jetnews/ui/home/PostCards.kt [hunk 22]
```diff
@@ -74,8 +77,8 @@ fun AuthorAndReadTime(
 
 @Composable
 fun PostImage(post: Post, modifier: Modifier = Modifier) {
-    Image(
-        painter = painterResource(post.imageThumbId),
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/ui/home/PostCards.kt:78]: Remove unused import for painterResource

File: gradle/libs.versions.toml [hunk 23]
```diff
@@ -57,6 +61,8 @@ secrets = "2.0.1"
 # @keep
 targetSdk = "33"
 version-catalog-update = "0.8.4"
+firebaseCrashlyticsBuildtools = "3.0.2"
```
Comment (jp-hoehmann) [inline gradle/libs.versions.toml:64]: Remove this

File: gradle/libs.versions.toml [hunk 24]
```diff
@@ -152,16 +159,23 @@ junit = { module = "junit:junit", version.ref = "junit" }
...
+ktor-client-okhttp = { module = "io.ktor:ktor-client-okhttp", version.ref = "ktor" }
+ktor-serialization-kotlinx-json = { module = "io.ktor:ktor-serialization-kotlinx-json", version.ref = "ktor" }
 play-services-wearable = { module = "com.google.android.gms:play-services-wearable", version.ref = "play-services-wearable" }
 robolectric = { module = "org.robolectric:robolectric", version.ref = "robolectric" }
 roborazzi = { module = "io.github.takahirom.roborazzi:roborazzi", version.ref = "roborazzi" }
 roborazzi-compose = { module = "io.github.takahirom.roborazzi:roborazzi-compose", version.ref = "roborazzi" }
 roborazzi-rule = { module = "io.github.takahirom.roborazzi:roborazzi-junit-rule", version.ref = "roborazzi" }
 rometools-modules = { module = "com.rometools:rome-modules", version.ref = "rome" }
 rometools-rome = { module = "com.rometools:rome", version.ref = "rome" }
+firebase-crashlytics-buildtools = { group = "com.google.firebase", name = "firebase-crashlytics-buildtools", version.ref = "firebaseCrashlyticsBuildtools" }
```
Comment (jp-hoehmann) [inline gradle/libs.versions.toml:177]: Remove this

File: app/src/main/java/com/example/jetnews/ui/home/PostCardTop.kt
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/ui/home/PostCardTop.kt:1]: Fix previews

File: app/src/main/java/com/example/jetnews/ui/home/HomeScreens.kt
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/ui/home/HomeScreens.kt:1]: Fix previews

File: app/src/main/java/com/example/jetnews/ui/article/ArticleScreen.kt
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/ui/article/ArticleScreen.kt:1]: Fix previews

**Review by jp-hoehmann — COMMENTED** [review]
Thanks for the PR, your changes look good for the most part :-)

There's three major things that need fixing:
* The app currently crashes when there is no internet connection, instead it should show an error to the user.
* The tests currently fail to build, because BlockingFakePostsRepository doesn't exist, instead the existing tests should still run and none of the tests should fail.
* The previews for a bunch of the Composables are either removed or broken, instead all previews that existed previously should still exist and be usable.

Other than that, there are some minor things I've commented on the individual files.

File: app/src/main/java/com/example/jetnews/data/posts/impl/HttpPostsRepository.kt (hunk 3 above)
Comment (mvlexs) [reply app/src/main/java/com/example/jetnews/data/posts/impl/HttpPostsRepository.kt]: Code is depending on imported Modules, therefore they should not be removed

**Comment by mvlexs** [comment]
applied changes to comply with response to PR-1, except for problem with crashes without internet, still working on an optimal solution

**Comment by mvlexs** [comment]
fixed: app crashing when no internet connection is available

File: app/src/main/java/com/example/jetnews/data/posts/impl/HttpPostsRepository.kt [hunk 25]
```diff
@@ -16,23 +16,35 @@
...
+import io.ktor.client.request.get
+import io.ktor.serialization.kotlinx.json.json
 import kotlinx.coroutines.Dispatchers
-import kotlinx.coroutines.delay
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.flow.update
 import kotlinx.coroutines.withContext
 
+
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/data/posts/impl/HttpPostsRepository.kt]: Unnecessary whitespace change.

File: app/src/main/java/com/example/jetnews/data/posts/impl/HttpPostsRepository.kt [hunk 26]
```diff
@@ -16,23 +16,35 @@
...
 /**
  * Implementation of PostsRepository that returns a hardcoded list of
  * posts with resources after some delay in a background thread.
  */
-class FakePostsRepository : PostsRepository {
+
+
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/data/posts/impl/HttpPostsRepository.kt]: Remove extra whitespace

File: app/src/main/java/com/example/jetnews/data/posts/impl/HttpPostsRepository.kt [hunk 27]
```diff
@@ -16,23 +16,35 @@
...
+class HttpPostsRepository : PostsRepository {
+
+
+
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/data/posts/impl/HttpPostsRepository.kt]: Remove extra whitespace

File: app/src/main/java/com/example/jetnews/data/posts/impl/HttpPostsRepository.kt [hunk 28]
```diff
@@ -42,25 +54,57 @@ class FakePostsRepository : PostsRepository {
...
+                val posts: PostsFeed = client.get(apiUrl).body()
+
                 postsFeed.update { posts }
                 Result.Success(posts)
+            } catch (e: Exception) {
+                Log.d("ErrorMalik",e.toString())
+                Result.Error(e)
             }
+
+
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/data/posts/impl/HttpPostsRepository.kt]: Remove extra whitespace.

File: app/src/main/java/com/example/jetnews/data/posts/impl/PostsData.kt [hunk 29]
```diff
@@ -416,7 +415,7 @@ val paragraphsPost2 = listOf(
         class MyVMFactory @Inject constructor(
           private val vmMap: Map<String, @JvmSuppressWildcards Provider<ViewModel>>
         ) { 
-            ... 
+        ...
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/data/posts/impl/PostsData.kt]: Unnecessary whitespace change.

**Review by jp-hoehmann — CHANGES_REQUESTED** [review]
Looks good. All working now. I have a few formatting nitpicks, but that's it.

File: app/src/main/java/com/example/jetnews/data/posts/impl/HttpPostsRepository.kt [hunk 30]
```diff
@@ -42,25 +54,57 @@ class FakePostsRepository : PostsRepository {
...
+                val client = HttpClient {
+
+                    install(ContentNegotiation){
+                        json()
+                    }
+
+                }
+
+                val posts: PostsFeed = client.get(apiUrl).body()
+                Log.d("DebugMalik",posts.toString())
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/data/posts/impl/HttpPostsRepository.kt]: Remove this.

File: app/src/main/java/com/example/jetnews/data/posts/impl/HttpPostsRepository.kt [hunk 31]
```diff
@@ -42,25 +54,57 @@ class FakePostsRepository : PostsRepository {
...
+
+                val post = posts.allPosts.find { it.id == postId }
+
+                if (post == null) {
+                    Result.Error(IllegalArgumentException("Post not found"))
+                } else {
+                    Result.Success(post)
+                }
+            } catch (e: Exception) {
+                Log.d("ErrorMalik",e.toString())
```
Comment (jp-hoehmann) [inline app/src/main/java/com/example/jetnews/data/posts/impl/HttpPostsRepository.kt]: Remove this.

**Comment by mvlexs** [comment]
removed unnecessary whitespace changes and logs

**Pull request merged**